src/input/iterator.rs

```diff
@@ -0,0 +1,196 @@
+            );
+            self.errors.push(err);
+        }
+        if self.errors.is_empty() {
+            Ok(())
+        } else {
+            Err(ValError::LineErrors(std::mem::take(&mut self.errors)))
+        }
+    }
+    fn check_max_length<I: Input<'data>>(
+        &self,
+        current_length: usize,
+        max_length: usize,
+        input: &'data I,
+    ) -> ValResult<'data, ()> {
+        if max_length < current_length {
+            return Err(ValError::new(
+                ErrorType::TooLong {
+                    field_type: self.field_type.to_string(),
+                    max_length,
+                    actual_length: current_length,
+                },
+                input,
+            ));
+        }
+        Ok(())
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+pub fn validate_infallible_iterator<'s, 'data, V, O, W, L>(
+    py: Python<'data>,
+    input: &'data impl Input<'data>,
+    extra: &'s Extra<'s>,
+    definitions: &'data Definitions<CombinedValidator>,
+    recursion_guard: &'s mut RecursionGuard,
+    checks: &mut IterableValidationChecks<'data>,
+    iter: impl Iterator<Item = &'data V>,
+    items_validator: &'s CombinedValidator,
+    output: &mut O,
+    write: &mut W,
+    len: &L,
+) -> ValResult<'data, ()>
+where
+    V: Input<'data> + 'data,
+    W: FnMut(&mut O, PyObject) -> PyResult<()>,
+    L: Fn(&O) -> usize,
+{
+    for (index, value) in iter.enumerate() {
+        let result = items_validator
+            .validate(py, value, extra, definitions, recursion_guard)
+            .map_err(|e| e.with_outer_location(index.into()));
+        if let Some(value) = checks.filter_validation_result(result, input)? {
+            write(output, value)?;
+            checks.check_output_length(len(output), input)?;
+        }
+    }
+    checks.finish(input)?;
+    Ok(())
+}
+
+#[allow(clippy::too_many_arguments)]
+pub fn validate_fallible_iterator<'s, 'data, V, O, W, L>(
+    py: Python<'data>,
+    input: &'data impl Input<'data>,
+    extra: &'s Extra<'s>,
+    definitions: &'data Definitions<CombinedValidator>,
+    recursion_guard: &'s mut RecursionGuard,
+    checks: &mut IterableValidationChecks<'data>,
+    iter: impl Iterator<Item = ValResult<'data, &'data V>>,
+    items_validator: &'s CombinedValidator,
+    output: &mut O,
+    write: &mut W,
+    len: &L,
+) -> ValResult<'data, ()>
+where
+    V: Input<'data> + 'data,
+    W: FnMut(&mut O, PyObject) -> PyResult<()>,
+    L: Fn(&O) -> usize,
+{
+    for (index, result) in iter.enumerate() {
+        let value = result?;
+        let result = items_validator
+            .validate(py, value, extra, definitions, recursion_guard)
+            .map_err(|e| e.with_outer_location(index.into()));
+        if let Some(value) = checks.filter_validation_result(result, input)? {
+            write(output, value)?;
+            checks.check_output_length(len(output), input)?;
+        }
+    }
+    checks.finish(input)?;
+    Ok(())
+}
```

Comment:
Splitting these up (same for the `dict.rs` versions) was an attempt to improve performance for the non-fallible case. It seemed to make a tiny difference, but to be honest it might be worth just collapsing into a single function to save a couple hundred LOC.